dbms/src/Interpreters/PredicateExpressionsOptimizer.cpp

```diff
@@ -121,15 +142,71 @@ bool PredicateExpressionsOptimizer::allowPushDown(const ASTSelectQuery * subquer
         {
             const auto & function = FunctionFactory::instance().tryGet(subquery_function->name, context);
 
-            /// Skip lambda、tuple and other special functions
+            /// Skip lambda, tuple and other special functions
             if (function && function->isStateful())
                 return false;
         }
+    }
+
+    const auto * ast_join = ast_select->join();
+
+    /// XXX: code depends on the order and availability of AST children
+    const auto * left_table_expr = ast_join ? ast_select->tables->as<ASTTablesInSelectQuery>()
+                                                  ->children[0]
+                                                  ->as<ASTTablesInSelectQueryElement>()
+                                                  ->table_expression->as<ASTTableExpression>()
+                                            : nullptr;
+    const auto * right_table_expr = ast_join ? ast_select->tables->as<ASTTablesInSelectQuery>()
+                                                   ->children[1]
+                                                   ->as<ASTTablesInSelectQueryElement>()
+                                                   ->table_expression->as<ASTTableExpression>()
+                                             : nullptr;
+    const auto * left_subquery = (left_table_expr && left_table_expr->subquery)
+        ? left_table_expr->subquery->children[0]->as<ASTSelectWithUnionQuery>()->list_of_selects->children[0]->as<ASTSelectQuery>()
+        : nullptr;
+    const auto * right_subquery = (right_table_expr && right_table_expr->subquery)
+        ? right_table_expr->subquery->children[0]->as<ASTSelectWithUnionQuery>()->list_of_selects->children[0]->as<ASTSelectQuery>()
+        : nullptr;
+
+    if (ast_join)
+    {
+        /// Check right side for LEFT'o'FULL JOIN
+        if (isLeftOrFull(ast_join->table_join->as<ASTTableJoin>()->kind) && right_subquery == subquery)
+        {
+            return false;
+        }
 
-        return true;
+        /// Check left side for RIGHT'o'FULL JOIN
+        if (isRightOrFull(ast_join->table_join->as<ASTTableJoin>()->kind) && left_subquery == subquery)
+        {
+            return false;
+        }
     }
 
-    return false;
+    for (const auto & [identifier, prefix] : dependencies)
```

Comment:
I've spent some time to extract this cycle into own func to be clear that it doesn't use more arguments that function gets. I don't think it's a good idea to mix it with other the code again cause one day it starts to use all the other variables.

Reply:
Fixed